lib/Controller/ContextController.php

```diff
@@ -0,0 +1,201 @@
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\DB\Exception;
+use OCP\IL10N;
+use OCP\IRequest;
+use Psr\Log\LoggerInterface;
+
+/**
+ * @psalm-import-type TablesContext from ResponseDefinitions
+ */
+
+class ContextController extends AOCSController {
+	private ContextService $contextService;
+
+	public function __construct(
+		IRequest        $request,
+		LoggerInterface $logger,
+		IL10N           $n,
+		string          $userId,
+		ContextService  $contextService
+	) {
+		parent::__construct($request, $logger, $n, $userId);
+		$this->contextService = $contextService;
+		$this->userId = $userId;
+	}
+
+	/**
+	 * [api v3] Get all contexts available to the requesting person
+	 *
+	 * Return an empty array if no contexts were found
+	 *
+	 * @return DataResponse<Http::STATUS_OK, TablesContext[], array{}>|DataResponse<Http::STATUS_INTERNAL_SERVER_ERROR, array{message: string}, array{}>
+	 *
+	 * 200: reporting in available contexts
+	 *
+	 * @NoAdminRequired
+	 */
+	public function index(): DataResponse {
+		try {
+			$contexts = $this->contextService->findAll($this->userId);
+			return new DataResponse($this->contextsToArray($contexts));
+		} catch (InternalError|Exception $e) {
+			return $this->handleError($e);
+		}
+	}
+
+	/**
+	 * [api v3] Get information about the requests context
+	 *
+	 * @return DataResponse<Http::STATUS_OK, TablesContext, array{}>|DataResponse<Http::STATUS_INTERNAL_SERVER_ERROR, array{message: string}, array{}>
+	 *
+	 * 200: returning the full context information
+	 * 404: context not found or not available anymore
+	 *
+	 * @NoAdminRequired
+	 */
+	public function show(int $contextId): DataResponse {
+		try {
+			$context = $this->contextService->findById($contextId, $this->userId);
+			return new DataResponse($context->jsonSerialize());
+		} catch (InternalError|Exception $e) {
+			return $this->handleError($e);
+		}
+	}
+
+	/**
+	 * [api v2] Create a new context and return it
+	 *
+	 * @NoAdminRequired
+	 *
+	 * @param string $name Name of the context
+	 * @param string $iconName Material design icon name of the context
+	 * @param string $description Descriptive text of the context
+	 * @param array $nodes optional nodes to be connected to this context
+	 *
+	 * @return DataResponse<Http::STATUS_OK, TablesContext, array{}>|DataResponse<Http::STATUS_INTERNAL_SERVER_ERROR, array{message: string}, array{}>
+	 *
+	 * 200: Tables returned
+	 */
+	public function create(string $name, string $iconName, string $description = '', array $nodes = []): DataResponse {
+		try {
+			return new DataResponse($this->contextService->create($name, $iconName, $description, $nodes, $this->userId, 0)->jsonSerialize());
+		} catch (Exception $e) {
+			return $this->handleError($e);
+		}
+	}
+
+	/**
+	 * @NoAdminRequired
+	 * @CanManageContext
+	 */
+	public function update(int $contextId, ?string $name, ?string $iconName, ?string $description): DataResponse {
```

Comment:
@enjeck the commit from right now adds support to provide node for the updates endpoint, just as it can be provided in create. 

I have not removed the dedicated endpoints, but if you confirm that we do not need them, I will remove them. Less ballast :)

Reply:
Ah, yes! That is because the *order* argument for the node is not provided. Actually, this came across one weekend's night (goodbye sleep). Should have pointed this out explicitly. 

On create we do not accept this parameter, but set it dynamically, based on the order the nodes were submitted. I considered this good enough on create (and as long as we just have that one page). But this is far from ideal. 

The other `addNode` endpoint also does not require the order, but if it is not given it hardcodes it 100. This is also marked as FIXME to either require this value, or calculate a value to just append it.

And then there is the dedicated endpoint to update a page's order.

Needs a better solution. 

Temporary quick fixes:
- on the FE, supply the order parameter for the node (cf. the example in the description)
- on the BE, i can hard code the 100 as well for now

And then think this aspect over again. What is your preferred way to update a node's order from the FE point of view?

Reply:
To add: on the update endpoint, we do not update the page content when only the order is changed of an existing node,

Reply:
> on the BE, i can hard code the 100 as well for now

just committed it for now